### PR TITLE
[servers] extract rpc modules to utils 

### PR DESCRIPTION
**Pull request by niklasad1**
Closing #307

One annoyance with this, is that it's possible to register subscription on HTTP server with this.
Maybe introduce some trait for circumvent this?!

### CODE REVIEW

File: utils/src/server/rpc_module.rs [hunk 1]
```diff
@@ -1,15 +1,29 @@
...
 use rustc_hash::FxHashMap;
 use serde::Serialize;
+use serde_json::value::to_raw_value;
 use std::sync::Arc;
 
+/// Method.
+pub type Method = Box<dyn Send + Sync + Fn(RpcId, RpcParams, RpcSender, ConnectionId) -> anyhow::Result<()>>;
+/// Methods registered.
+pub type Methods = FxHashMap<&'static str, Method>;
+/// Connection ID.
```
Comment (dvdplm) [inline utils/src/server/rpc_module.rs:20]:
```suggestion
/// Connection ID. Set to `0` for http and to the connection ID for websockets.
```

File: utils/src/server/rpc_module.rs [hunk 2]
```diff
@@ -1,15 +1,29 @@
...
+pub type ConnectionId = usize;
+/// Subscription ID.
+pub type SubscriptionId = u64;
+type Subscribers = Arc<Mutex<FxHashMap<(ConnectionId, SubscriptionId), mpsc::UnboundedSender<String>>>>;
+
+/// Abstract JSON-RPC module that be registered on server or merged with other modules.
```
Comment (dvdplm) [inline utils/src/server/rpc_module.rs:26]:
```suggestion
/// Sets of JSON-RPC methods can be organized into "module" that are in turn registered on server or, alternatively, merged with other modules to construct a cohesive API.
```

File: utils/src/server/rpc_module.rs [hunk 3]
```diff
@@ -122,11 +136,13 @@ impl RpcModule {
...
 	}
 
-	pub(crate) fn into_methods(self) -> Methods {
+	/// Convert a module into methods.
+	pub fn into_methods(self) -> Methods {
 		self.methods
 	}
 
-	pub(crate) fn merge(&mut self, other: RpcModule) -> Result<(), Error> {
+	/// Merge modules.
```
Comment (dvdplm) [inline utils/src/server/rpc_module.rs:144]:
```suggestion
	/// Merge two [`RpcModules`] by adding all [`Method`]s from `other` into `self`. Fails if any of the methods in `other` is present already.
```

File: utils/src/server/rpc_module.rs [hunk 4]
```diff
@@ -188,3 +206,42 @@ impl<Context> RpcContextModule<Context> {
 		self.module
 	}
 }
+
+/// The sending end of registered subscription.
```
Comment (dvdplm) [inline utils/src/server/rpc_module.rs:210]:
```suggestion
/// Used by the server to send data back to subscribers.
```

File: utils/src/server/rpc_module.rs [hunk 5]
```diff
@@ -188,3 +206,42 @@ impl<Context> RpcContextModule<Context> {
...
+#[derive(Clone)]
+pub struct SubscriptionSink {
+	method: &'static str,
+	subscribers: Subscribers,
+}
+
+impl SubscriptionSink {
+	/// Send.
```
Comment (dvdplm) [inline utils/src/server/rpc_module.rs:218]:
```suggestion
	/// Send data back to subscribers. If a send fails (likely a broken connection) the subscriber is removed from the sink. O(n) in the number of subscribers.
```

**Review by dvdplm — COMMENTED** [review]
Added some more docs.

**Comment by dvdplm** [comment]
> One annoyance with this, is that it's possible to register subscription on HTTP server with this.

Seems like a minor annoyance. Let's log a low-prio issue and move on.

File: utils/src/server/mod.rs [hunk 6]
```diff
@@ -0,0 +1,11 @@
+//! Shared modules for the JSON-RPC servers.
+
+/// Helpers.
+pub mod helpers;
+/// Abstract JSON-RPC modules that can be used to register methods on a server.
+pub mod rpc_module;
+
+/// Sender.
+pub type RpcSender<'a> = &'a futures_channel::mpsc::UnboundedSender<String>;
+/// RPC ID.
```
[the remark this answers is not in the document]
Comment (niklasad1) [reply utils/src/server/mod.rs:10]: yes and yes

File: utils/src/server/mod.rs [hunk 7]
```diff
@@ -0,0 +1,11 @@
+//! Shared modules for the JSON-RPC servers.
+
+/// Helpers.
+pub mod helpers;
+/// Abstract JSON-RPC modules that can be used to register methods on a server.
+pub mod rpc_module;
+
+/// Sender.
```
[the remark this answers is not in the document]
Comment (niklasad1) [reply utils/src/server/mod.rs]: renamed it to `MethodSink` and added some docs, should be more understandable now.

Can you take another look?

File: utils/src/server/helpers.rs [hunk 8]
```diff
@@ -1,27 +1,12 @@
...
-/// RPC ID.
-pub type RpcId<'a> = Option<&'a RawValue>;
-/// Method registered in the server.
-pub type Method = Box<dyn Send + Sync + Fn(RpcId, RpcParams, RpcSender, ConnectionId) -> anyhow::Result<()>>;
-/// Methods registered in the Server.
-pub type Methods = FxHashMap<&'static str, Method>;
 
 /// Helper for sending JSON-RPC responses to the client
-pub fn send_response(id: RpcId, tx: RpcSender, result: impl Serialize) {
+pub fn send_response(id: JsonRpcRawId, tx: &mpsc::UnboundedSender<String>, result: impl Serialize) {
```
Comment (dvdplm) [inline utils/src/server/helpers.rs:9]:
```suggestion
pub fn send_response(id: JsonRpcRawId, tx: &MethodSink, result: impl Serialize) {
```
?

File: utils/src/server/helpers.rs [hunk 9]
```diff
@@ -37,7 +22,7 @@ pub fn send_response(id: RpcId, tx: RpcSender, result: impl Serialize) {
 }
 
 /// Helper for sending JSON-RPC errors to the client
-pub fn send_error(id: RpcId, tx: RpcSender, error: JsonRpcErrorObject) {
+pub fn send_error(id: JsonRpcRawId, tx: &mpsc::UnboundedSender<String>, error: JsonRpcErrorObject) {
```
Comment (dvdplm) [inline utils/src/server/helpers.rs:25]: same?

File: utils/src/server/mod.rs [hunk 10]
```diff
@@ -0,0 +1,6 @@
+//! Shared modules for the JSON-RPC servers.
+
+/// Helpers.
+pub mod helpers;
+/// JSON-RPC "modules" groups sets of methods that belong together and handles method/subscription registration.
```
Comment (dvdplm) [inline utils/src/server/mod.rs:5]:
```suggestion
/// JSON-RPC "modules" group sets of methods that belong together and handles method/subscription registration.
```

File: utils/src/server/rpc_module.rs [hunk 11]
```diff
@@ -1,15 +1,36 @@
...
 
+/// A `Method` is an RPC endpoint, callable with a standard JSON-RPC request,
+/// implemented as a function pointer to a `Fn` function taking four arguments:
+/// the `id`, `params`, a channel the function uses to communicate the result (or error)
+/// back to `jsonrpsee`, and the connection ID (useful for the websocket transport).
+pub type Method = Box<dyn Send + Sync + Fn(JsonRpcRawId, RpcParams, &MethodSink, ConnectionId) -> anyhow::Result<()>>;
+/// A collection of registered [`Method`]s.
+pub type Methods = FxHashMap<&'static str, Method>;
+/// Connection ID, used for stateful protocol such as WebSockets.
+/// For stateless protocols it's unused, so feel free to set it some hardcoded value.
```
Comment (dvdplm) [inline utils/src/server/rpc_module.rs:23]:
```suggestion
/// For stateless protocols such as http it's unused, so feel free to set it some hardcoded value.
```

File: utils/src/server/rpc_module.rs [hunk 12]
```diff
@@ -1,15 +1,36 @@
...
+pub type ConnectionId = usize;
+/// Subscription ID.
+pub type SubscriptionId = u64;
+/// Sink that is used to send back the result to the server for a specific method.
+pub type MethodSink = mpsc::UnboundedSender<String>;
+
+type Subscribers = Arc<Mutex<FxHashMap<(ConnectionId, SubscriptionId), MethodSink>>>;
+
+/// Sets of JSON-RPC methods can be organized into "module" that are in turn registered on server or,
```
Comment (dvdplm) [inline utils/src/server/rpc_module.rs:32]:
```suggestion
/// Sets of JSON-RPC methods can be organized into a "module" that are in turn registered on server or,
```

File: utils/src/server/rpc_module.rs [hunk 13]
```diff
@@ -139,6 +163,8 @@ impl RpcModule {
 	}
 }
 
+/// Similar to [`RpcModule`] but it wraps an additional context argument that can be used
```
Comment (dvdplm) [inline utils/src/server/rpc_module.rs:166]:
```suggestion
/// Similar to [`RpcModule`] but wraps an additional context argument that can be used
```

**Review by dvdplm — APPROVED** [review]

**Pull request merged**